R/map.R

```diff
@@ -112,38 +112,44 @@
 #'   map(summary) |>
 #'   map_dbl("r.squared")
 map <- function(.x, .f, ..., .progress = FALSE) {
-  .f <- as_mapper(.f, ...)
-  .Call(map_impl, environment(), ".x", ".f", "list", .progress)
+  map_("list", .x, .f, ..., .progress = .progress)
 }
 
 #' @rdname map
 #' @export
 map_lgl <- function(.x, .f, ..., .progress = FALSE) {
-  .f <- as_mapper(.f, ...)
-  .Call(map_impl, environment(), ".x", ".f", "logical", .progress)
+  map_("logical", .x, .f, ..., .progress = .progress)
 }
 
 #' @rdname map
 #' @export
 map_int <- function(.x, .f, ..., .progress = FALSE) {
-  .f <- as_mapper(.f, ...)
-  .Call(map_impl, environment(), ".x", ".f", "integer", .progress)
+  map_("integer", .x, .f, ..., .progress = .progress)
 }
 
 #' @rdname map
 #' @export
 map_dbl <- function(.x, .f, ..., .progress = FALSE) {
-  .f <- as_mapper(.f, ...)
-  .Call(map_impl, environment(), ".x", ".f", "double", .progress)
+  map_("double", .x, .f, ..., .progress = .progress)
 }
 
 #' @rdname map
 #' @export
 map_chr <- function(.x, .f, ..., .progress = FALSE) {
+  map_("character", .x, .f, ..., .progress = .progress)
+}
+
+map_ <- function(.type, .x, .f, ..., .progress = FALSE, .error_call = caller_env()) {
```

Comment:
A side effect of having `.error_call` here is that now functions that pass `.error_call` through the dots won't be able to pass that error call all the way through to `.f`, it will get partially matched here.

A bunch of vctrs functions are getting an `.error_call` arg

Same with `.type` and `.progress` I guess

``` r
library(purrr)

error_call <- current_env()

# `error_call` is passed through to `vec_rep2()`
vec_rep2 <- function(x, times, error_call = current_env()) {
  vec_rep(x, times, call = error_call)
}
map(1:5, vec_rep2, times = -1, error_call = error_call)
#> Error in `map()`:
#> ! In index 1.
#> Caused by error:
#> ! `times` must be a positive number.

# `.error_call` is matched by `map_()` and doesn't make it
# to `vec_rep2()`
vec_rep2 <- function(x, times, .error_call = current_env()) {
  vec_rep(x, times, call = .error_call)
}
map(1:5, vec_rep2, times = -1, .error_call = error_call)
#> Error:
#> ! In index 1.
#> Caused by error in `.f()`:
#> ! `times` must be a positive number.
```

Reply:
Hmmm, probably worth making this `..error_call`, but generally this is why using `map(1:5, \(x) vec_rep2(x, times = -1, .error_call = error_call))` is a good idea.